aws/resource_aws_kms_external_key.go

```diff
@@ -0,0 +1,481 @@
+	"github.com/hashicorp/terraform/helper/structure"
+	"github.com/hashicorp/terraform/helper/validation"
+)
+
+func resourceAwsKmsExternalKey() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsKmsExternalKeyCreate,
+		Read:   resourceAwsKmsExternalKeyRead,
+		Update: resourceAwsKmsExternalKeyUpdate,
+		Delete: resourceAwsKmsExternalKeyDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"deletion_window_in_days": {
+				Type:         schema.TypeInt,
+				Optional:     true,
+				Default:      30,
+				ValidateFunc: validation.IntBetween(7, 30),
+			},
+			"description": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringLenBetween(0, 8192),
+			},
+			"enabled": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Computed: true,
+			},
+			"expiration_model": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"key_material_base64": {
+				Type:      schema.TypeString,
+				Optional:  true,
+				ForceNew:  true,
+				Sensitive: true,
+			},
+			"key_state": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"key_usage": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"policy": {
+				Type:             schema.TypeString,
+				Optional:         true,
+				Computed:         true,
+				DiffSuppressFunc: suppressEquivalentAwsPolicyDiffs,
+				ValidateFunc: validation.All(
+					validation.StringLenBetween(0, 32768),
+					validation.ValidateJsonString,
+				),
+			},
+			"tags": tagsSchema(),
+			"valid_to": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.ValidateRFC3339TimeString,
+			},
+		},
+	}
+}
+
+func resourceAwsKmsExternalKeyCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).kmsconn
+
+	input := &kms.CreateKeyInput{
+		KeyUsage: aws.String(kms.KeyUsageTypeEncryptDecrypt),
+		Origin:   aws.String(kms.OriginTypeExternal),
+	}
+
+	if v, ok := d.GetOk("description"); ok {
+		input.Description = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("policy"); ok {
+		input.Policy = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("tags"); ok {
+		input.Tags = tagsFromMapKMS(v.(map[string]interface{}))
+	}
+
+	var output *kms.CreateKeyOutput
+	// AWS requires any principal in the policy to exist before the key is created.
+	// The KMS service's awareness of principals is limited by "eventual consistency".
+	// They acknowledge this here:
+	// http://docs.aws.amazon.com/kms/latest/APIReference/API_CreateKey.html
+	err := resource.Retry(1*time.Minute, func() *resource.RetryError {
+		var err error
+
+		output, err = conn.CreateKey(input)
+
+		if isAWSErr(err, kms.ErrCodeMalformedPolicyDocumentException, "") {
+			return resource.RetryableError(err)
```

Comment:
Since IAM changes such as new IAM principals can take some time to propagate due to that service's eventual consistency, KMS will report this error until it can validate the policy itself. We offer this reasonable amount of time to retry for the eventual consistency case but still error out after in case the policy is indeed correct. For reference, this logic was copied from the existing `aws_kms_key` resource creation logic. I can move the more descriptive code comment done to this conditional if you think it will help.

Reply:
Thanks for the clarification :+1:

I think moving the comment down would help make the relation between the `kms.ErrCodeMalformedPolicyDocumentException` and the comment. I also recommend adding `KMS will report this error until it can validate the policy itself` to the comment to drive the point home.